src/provider.tsx

```diff
@@ -0,0 +1,29 @@
+'use client';
+
+import * as React from 'react';
+import { checkSessionAction } from './actions.js';
+
+export const Provider = ({ children }: React.PropsWithChildren) => {
+  React.useEffect(() => {
+    const handleVisibilityChange = async () => {
+      if (document.visibilityState === 'visible') {
+        try {
+          const hasSession = await checkSessionAction();
+          if (!hasSession) {
+            throw new Error('Session expired');
+          }
+        } catch (error) {
+          window.location.reload();
+        }
+      }
+    };
+
+    window.addEventListener('visibilitychange', handleVisibilityChange);
```

Comment:
We should leave a comment on why we're doing this and the edge case we're trying to solve.